contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,193 @@
+# Out-of-tree client authentication providers
+
+Author: @ericchiang
+
+# Objective
+
+This document describes a credential rotation strategy for kubectl and client-go
+using an exec-based plugin mechanism.
+
+# Background
+
+Kubernetes clients can provide three kind of credentials: bearer tokens, TLS
+client certs, and basic auth username and password. Kubeconfigs can either
+in-line the credential, load credentials from a file, or can use an `AuthProvider`
+to actively fetch and rotate credentials. `AuthProviders` are compiled into kubectl
+and target specific providers (GCP, Keystone, Azure AD) or implement a
+specification supported but a subset of vendors (OpenID Connect).
+
+External tools that do rotation generally have to be called on a regular bases in
+case the current credentials have expired, which leads to wrapping `kubectl` with
+another tool or aliasing. For example, this is a [real example](
+https://github.com/heptio/authenticator#4-set-up-kubectl-to-use-heptio-authenticator-for-aws-tokens)
+from Heptio's AWS authenticator:
+
+```
+kubectl --kubeconfig /path/to/kubeconfig --token "$(heptio-authenticator-aws token -i CLUSTER_ID -r ROLE_ARN)" [...]
+```
+
+Beside resulting in a long command, this potentially encourages distributions to
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
+kubectl (and client-go) should be able to receive credentials by executing a
+command and reading that command's stdout.
+
```

Comment:
The format of the output must be defined to ensure interoperability between kubectl/client-go and the external authenticator. I guess a JSON schema with a version would be sufficient. This could prevent cases when the authentication could be broken by a change on the kubectl/client-go side.

Reply:
Alternatively we support jsonpath and integrator define their own scheme. I'm not sure if it's better but we do this in the gcp plugin today: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go#L77-L84

Reply:
I think @cosmincojocar is talking about using a structured, versioned API so we can ask for additional data beside the credentials in the future.

Do we really need versioned output? Do we expect the tool to pass things like expiry to us? I expected just printing the bearer token to be sufficient. We expect some of these tools to not know anything about kubectl.

If we don't go with a structured stdout API, I'd be in favor of some way of parsing the output like jsonpath.

Reply:
Or maybe get directly in the output a base64 encoded JWT.

Reply:
Do you plan to cache the token on kubectl/client-go side and invoke the external authenticator only when the token is about to expire? Or do you want to invoke the external authenticator every time a token is required?

I think, the authenticator can persist its tokens in a file after the initial authentication, especially in the case when the refresh token is supported. It can execute later on only the refresh flow when the token is close to the expiration.

I suspect that there are authenticators which do not support the refresh flow (e.g. client credentials). In that case it will request a new token from IdP instead of refresh it. Also in this case a sort of caching will be helpful in order to avoid overwhelming the IdP.

I know that these details are up to the authenticator implementation, but they will have an impact in the kubectl/client-go performance and reliability.

Reply:
For kubectl, I'd rather leave it up to the external authenticator to handle persisting tokens in a file. Some might not want the token to be saved to disk or need other special handling for their credentials.

There's room for better cache/refresh handling in general but I don't think it needs to be directly coupled to this change.

Reply:
Added some comments on why we went for an unversioned output.